### ISSUE TITLE
tc39 proposal pattern matching into userland

### ISSUE THREAD

**Issue by Pyrolistical** [samuelgoto/proposal-block-params#45]
I really like block params as it would generalize the syntax https://github.com/tc39/proposal-pattern-matching is trying to add.

Which happens to what I incidentally did using with [patcom](https://github.com/concept-not-found/patcom#comparision-of-tc39-pattern-matching-proposal-on-left-to-patcom-on-right)

[See example](https://github.com/samuelgoto/proposal-block-params/issues/45#issuecomment-1066171236)
~~We could probably write a tc39 proposal pattern matching polyfill using a block params polyfill + patcom.~~

**Comment by ljharb** [comment]
What about block params would work well with pattern matching?

**Comment by Pyrolistical** [comment]
Sorry, I could have provided an example.

This tc39 proposal pattern matching syntax
```js
when ({ status: 200, body, ...rest }):
  handleData(body, rest)
```

is the like [proposal block params do expression](https://github.com/samuelgoto/proposal-block-params#bindings) (note the first `rest` is now a matcher, not destructuring)
```js
when ({ status: 200, body, rest }) do (body, rest) {
  return handleData(body, rest)
}
```

which is implementable in patcom today as
```js
when ({ status: 200, body, rest }, (body, rest) => {
  return handleData(body, rest)
})
```

It wouldn't be possible to write a tc39 proposal pattern matching polyfill. I take that back. But you could get very close to the intent.

**Comment by ljharb** [comment]
Sure, i see how with block params and a `when` function you could simulate parts of it - but you could do the same with a callback function, no?

**Comment by hax** [comment]
Yes, you can do same with callback, but the motivation of this proposal it to provide better syntax for callback/blocks.

Note this proposal also include a nest block feature, so in pattern match case, it should be possible to write 
```js
match (value) do {
  ::when ({ status: 200, body: defined, rest }) do (body, rest) {
    return handleData(body, rest)
  }
}
```

Actually it's a little bit like Raku pattern match, you could execute any statements in match block.
